openhcl/openhcl_boot/src/main.rs

```diff
@@ -605,18 +605,23 @@ fn shim_main(shim_params_raw_offset: isize) -> ! {
         )
         .vtl0_alias_map_available()
     {
-        // On ARM, use the physical address width reported via device tree since
-        // it is not reliably reported architecturally. Since the position of
-        // the alias map bit is inferred from address size, the alias map is
-        // broken when the PA size is wrong.
-        if !cfg!(target_arch = "aarch64") {
-            let vtl0_alias_map = 1 << (arch::physical_address_bits(p.isolation_type) - 1);
-            if let Some(dt_vtl0_alias_map) = partition_info.vtl0_alias_map {
-                // if we got the alias map via device tree, it should match.
-                assert_eq!(dt_vtl0_alias_map, vtl0_alias_map);
-            }
-            partition_info.vtl0_alias_map = Some(vtl0_alias_map);
+        // If the vtl0 alias map was not provided in the devicetree, attempt to
+        // derive it from the architectural physical address bits.
+        //
+        // The value in the ID_AA64MMFR0_EL1 register used to determine the
+        // physical address bits can only represent multiples of 4. As a result,
+        // the Surface Pro X (and systems with similar CPUs) cannot properly
+        // report their address width of 39 bits. This causes the calculated
+        // alias map to be incorrect, which results in panics when trying to
+        // read memory and getting invalid data.
+        if partition_info.vtl0_alias_map.is_none() {
```

Comment:
I'd still make this x86 only since it's almost always going to be wrong on ARM.

Reply:
What's the scenario where we need to support guest VSM on an ARM host that doesn't have this fix?

Reply:
Offline, Trevor convinced me that there are cases where this approximation works on ARM, and we want to support those cases even with an old host that doesn't provide the precise alias map offset.